File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/mappers/MappersExposeSingleInvoke.kt
package com.perrystreet.woof.konsist.mappers

import com.lemonappdev.konsist.api.ext.list.modifierprovider.withPublicOrDefaultModifier
import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class MappersExposeSingleInvoke : BehaviorSpec() {
    init {
        Given("A DTOToDomain or DomainToUIModel mapper") {
            val mappers = KonsistUtils.mappers

            Then("It is a @Factory class with a single operator fun invoke") {
                mappers.assertTrue(message = Message) { mapper ->
                    val publicFunctions = mapper.functions().withPublicOrDefaultModifier()
                    mapper.hasAnnotationWithName("Factory") &&
                        publicFunctions.size == 1 &&
                        publicFunctions.single().let { it.name == "invoke" && it.hasOperatorModifier }
                }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "Mappers are @Factory classes exposing exactly one operator fun invoke().",
            why = "A mapper converts one type into another and nothing else. Injecting it as a class (not an object) keeps nested mappers composable.",
            howToFix = "Annotate with @Factory, name the conversion function invoke, and split extra conversions into their own mappers.",
            badExample = "object DogMapper { fun toDomain(dto: DogDTO): Dog; fun toUIModel(dog: Dog): DogCellUIModel }",
            goodExample = "@Factory\nclass DogDTOToDomainMapper { operator fun invoke(dto: DogDTO): Dog }",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/Mappers/MappersExposeSingleInvoke.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class MappersExposeSingleInvoke: QuickSpec {
    override class func spec() {
        Given("A DTOToDomain or DomainToUIModel mapper") {
            let mappers = WoofHarmonize.mappers

            Then("It is a @Factory class with a single callAsFunction") {
                mappers.assertTrue(rule: rule) { mapper in
                    let publicFunctions = mapper.functions.withoutModifier(.private)
                    return mapper.hasAttribute(named: "@Factory")
                        && publicFunctions.count == 1
                        && publicFunctions[0].name == "callAsFunction"
                }
            }
        }
    }

    private static let rule = Rule(
        description: "Mappers are @Factory classes exposing exactly one callAsFunction().",
        rationale: "A mapper converts one type into another and nothing else. Injecting it as a class (not a static helper) keeps nested mappers composable.",
        fixHint: "Annotate with @Factory, name the conversion function callAsFunction, and split extra conversions into their own mappers.",
        badExample: "enum DogMapper { static func toDomain(_ dto: DogDTO) -> Dog; static func toUIModel(_ dog: Dog) -> DogCellUIModel }",
        goodExample: "@Factory\nfinal class DogDTOToDomainMapper { func callAsFunction(_ dto: DogDTO) -> Dog }"
    )
}
